### ISSUE TITLE
Does nunjucks 2.* support calling methods on context objects?

### ISSUE THREAD

**Issue by dschissler** [mozilla/nunjucks#560]
With nunjucks 1.3 I was able to do the following: `{{ model.getName() }}` but with 2.0 I'm unable to do this.

**Comment by carljm** [comment]
Yes, nunjucks 2 supports that just fine. Something else has changed, maybe such that `model` is no longer the thing you thought it was? It could be related to some other change in nunjucks, though nothing comes to mind. But calling methods on context objects definitely still works.

When you say "I'm unable to do this," what actual error do you get? Can you provide a minimal template and context to reproduce the problem?

**Comment by dschissler** [comment]
I'm using Ampersand.js.

I'm able to get nunjucks to access the data by first converting it into a POJO.  So my model data (the data that is wrapped up) is what I think that it is.

I think that this issue may have to do with the member being a getter.  The view object is passed in and it has a getter named 'model' and the model has getters on it as well.  I'm guessing that `[]` access doesn't invoke the getter.

I could work around this by creating a few filters that can grab this data but its not ideal.

```
Template render error: (unknown path) [Line 2, Column 24]
  Error: Unable to call `model["get"]`, which is undefined or falsey
Stack trace:
nunjucks</exports.prettifyError@https://dev.org1.secretprojectname.com/js/entries/app.site.js:33336:1
nunjucks</Template<.render/<@https://dev.org1.secretprojectname.com/js/entries/app.site.js:34056:1
module.exports/obj.root/<@https://dev.org1.secretprojectname.com/js/entries/app.site.js:35811:14
root@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36851:4
module.exports/obj.root@https://dev.org1.secretprojectname.com/js/entries/app.site.js:35809:10
nunjucks</Template<.render@https://dev.org1.secretprojectname.com/js/entries/app.site.js:34049:1
exports.default@https://dev.org1.secretprojectname.com/js/entries/app.site.js:32881:15
renderWithTemplate@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30046:18
render@https://dev.org1.secretprojectname.com/js/entries/app.site.js:37542:6
._setRender/<.set/this._render@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30481:22
._createViewForModel@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30566:10
._getOrCreateByModel@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30570:48
._addViewForModel@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30581:21
Collection.prototype[method]@https://dev.org1.secretprojectname.com/js/entries/app.site.js:22319:17
._renderAll@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30643:10
.render@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30549:10
.renderCollection@https://dev.org1.secretprojectname.com/js/entries/app.site.js:30470:10
_useData@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36668:27
setData@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36266:8
setData@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36273:12
setData@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36210:8
setData@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36148:8
start@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36119:8
start@https://dev.org1.secretprojectname.com/js/entries/app.site.js:36458:8
screenChange@https://dev.org1.secretprojectname.com/js/entries/app.site.js:29901:1
_screenChange@https://dev.org1.secretprojectname.com/js/entries/app.site.js:514:1
Radio._callHandler@https://dev.org1.secretprojectname.com/js/entries/app.site.js:2290:1
request@https://dev.org1.secretprojectname.com/js/entries/app.site.js:2408:17
globalRequest@https://dev.org1.secretprojectname.com/js/entries/app.site.js:2195:13
executeScreenApi$@https://dev.org1.secretprojectname.com/js/entries/app.site.js:665:14
tryCatch@https://dev.org1.secretprojectname.com/js/entries/app.site.js:1243:38
invoke@https://dev.org1.secretprojectname.com/js/entries/app.site.js:1491:23
defineIteratorMethods/</prototype[method]@https://dev.org1.secretprojectname.com/js/entries/app.site.js:1276:17
invoke@https://dev.org1.secretprojectname.com/js/entries/app.site.js:1312:21
tryCatcher@https://dev.org1.secretprojectname.com/js/commons/app.js:7198:17
[23]</module.exports/Promise.prototype._settlePromiseFromHandler@https://dev.org1.secretprojectname.com/js/commons/app.js:5304:14
[23]</module.exports/Promise.prototype._settlePromiseAt@https://dev.org1.secretprojectname.com/js/commons/app.js:5378:14
[23]</module.exports/Promise.prototype._settlePromises@https://dev.org1.secretprojectname.com/js/commons/app.js:5494:10
[2]</Async.prototype._drainQueue@https://dev.org1.secretprojectname.com/js/commons/app.js:2784:14
[2]</Async.prototype._drainQueues@https://dev.org1.secretprojectname.com/js/commons/app.js:2794:6
Async/this.drainQueues@https://dev.org1.secretprojectname.com/js/commons/app.js:2676:10
```

**Comment by dschissler** [comment]
Perhaps an issue should be brought up to support getters?

**Comment by carljm** [comment]
Yes, possibly -- especially if they really did work on 1.3. We don't need a new issue, you can just re-title this one. If the issue really is getters, it would be great to confirm that with a minimal reproduction case (sending an object with a getter into a template context) that doesn't rely on any other external frameworks like Ampersand. Even better if that repro case is provided as a new test case in the nunjucks test suite :-)

**Comment by dschissler** [comment]
Can you refer me to a bootstrap environment to get nunjucks working?

I suspect that nunjucks has gone the direction of parsing the `{{}}` content and then using that data to exclusively use bracket notation.  That seems like the dot notation would be better but then one would have to worry about illegal characters in the property name.  So there likely needs to be an extra check there if it is a property.  I don't know how to use bracket notation to invoke a getter but it sounds like perfect research for me after I get home from beers.

**Comment by carljm** [comment]
I think actually the simplest bootstrap environment (and the one that moves us closest to a fix, too) is just adding a test to the nunjucks test suite. If you look in `tests/compiler.js` you'll see a bunch of existing end-to-end tests, and it should be pretty obvious how to add a new one. You can define a template as a string, provide a context, and then assert what the output should be.

To run the tests, you'll need to `npm install` and then `npm test`. Feel free to push even just a failing test as a pull request, though if you're able to push on and find and implement a fix, even better!

The behavior of a dot attribute access in the template is defined in the `memberLookup` function in `runtime.js`, and your hypothesis is correct that nunjucks always uses bracket notation internally. (That didn't change between 1.3 and 2.0 though, so I'm still a bit mystified by your assertion that this worked in 1.3 -- you sure nothing else changed in your code along with the nunjucks upgrade?)

**Comment by dschissler** [comment]
> -- you sure nothing else changed in your code along with the nunjucks upgrade?)

I'm using nunjucks-loader, which went to 2.0 with the nunjucks 2.0 release, but I really can't imagine in an ideal that they managed to devolve into mutating context properties just for fun.

Some things did change between 1.3 and 2.0 though, like now I need to `export` macros instead of `import`ing them.  So for a multi-tab display I end up now with a quick-fix looking like `tab.tab(...)`.

**Comment by carljm** [comment]
> Some things did change between 1.3 and 2.0 though, like now I need to `export` macros instead of `import` ing them. So for a multi-tab display I end up now with a quick-fix looking like tab.tab(...).

Now I'm really confused. Nunjucks doesn't even have an `export` tag or keyword, AFAIK, so I have no idea what you mean by "now I need to export macros instead of importing them." And I don't think anything significant changed from 1.3 to 2.0 in terms of how `import` works.

**Comment by dschissler** [comment]
Sorry, I meant `import` instead of `include`

**Comment by dschissler** [comment]
The following quick test works with properties:

```javascript
'use strict';
let nunjucks = require('nunjucks');

let html;

let o1 = {
  _value: 'abc',
  get value() {
    return this._value;
  }
}
// Ouputs 'abc'
console.log(o1.value);
html = nunjucks.renderString('{{o.value}}', {o: o1});
// Ouputs 'abc'
console.log(html);

let O2 = function() {
  this._value = 'abc';
  Object.defineProperty(this, 'value', {
    get: function() { return this._value }
  });
}
let o2 = new O2();
html = nunjucks.renderString('{{o.value}}', {o: o2});
// Ouputs 'abc'
console.log(html);
```

So I think that the issue is somehow in `ampersand-state` or `nunjucks-loader`.

**Comment by carljm** [comment]
Ok, thanks for exploring!

**Comment by dschissler** [comment]
Reopen.

I found some strange breaking behavior.  Perhaps I don't understand how to use Object.defineProperty correctly but this is similar way to how `ampersand-state` uses it.  Nunjucks@2 doesn't like getters when it was defined afterwards, maybe it is detecting some different in prototype and treating it differently


```javascript
'use strict';
let nunjucks = require('nunjucks');
let model, view, html;

let Model = function() {
  this._modelValue = 'model_value';
  Object.defineProperty(this, 'value', {
    get: function() { return this._modelValue }
  });
}


model = new Model();
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// This doesn't work
// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
html = nunjucks.renderString('{{ value }}', model);
console.log(html);



// Try again with a POJO
model = {
  _modelValue: 'model_value',
  get value() {
    return this._modelValue;
  }
}

// This direct propery access works
html = nunjucks.renderString('{{ _modelValue }}', model);
console.log(html);

// This getter access works
html = nunjucks.renderString('{{ value }}', model);
console.log(html);
```

**Comment by carljm** [comment]
Ok, seems like the problem is not in the code to access attributes of context objects (the `memberLookup` function I mentioned above), it's in the code for accessing top-level attributes of the context.

**Comment by dschissler** [comment]
I'm guessing that `memberLookup` is some sort of parser and the "top-level" accessing part handles the javascript calls?

**Comment by carljm** [comment]
I'm not sure what you mean by that. Nunjucks compiles templates (using a lexer, a parser, and a compiler, as usual). So everything in the template is parsed by nunjucks; nothing is ever passed through as raw JS code or anything like that.

`runtime.memberLookup` is the JS function used in the compiler-generated JS code when it sees a dotted attribute access in the parsed template AST. I just checked to refresh my memory, and `runtime.contextOrFrameLookup` is the runtime function used to handle a top-level name lookup. So that's probably where you'd want to start looking to debug this (probably `Context.lookup`, since that's where `contextOrFrameLookup` will end up if the name isn't in a local frame, which in your test example it isn't).

Actually now that I look at `Context` (in `environment.js`) it's pretty clear where the issue probably lies. It's in `Context.init`, where the context is duplicated using `for (var k in ctx) { if (ctx.hasOwnProperty(k)) { ... } }`. I presume that doesn't handle defined properties well.

Looks like that copying of the context was introduced fairly recently in https://github.com/mozilla/nunjucks/commit/f99bcea in order to fix issues with context leakage in imports. That explains why this is a regression in 2.0. So I guess the question is, can that context-copying code be adapted in order to properly support defined properties?

**Comment by dschissler** [comment]
I think that I can work around this for now by passing in the model as the context instead of the view.  Then I can access the model properties with the actual "get()" function and then later when this issue is resolved I'll change the templates back to use the property getter.

For a case of why this is important: it will really be nice to have getter support because then I can abstract away the source of the values such as in the case of some record properties calling `{{ get('active') }}` and then for the name which is stored in another table in a `nedb` JS database it is accessed with `{{ getName() }}`.  It would be great to be able to just write: `{{ active}}` or `{{ name }}` and to have the model handle where the data comes from.

**Comment by dschissler** [comment]
I just read your last comment.  I don't think that I'm well qualified to work on the insides of a template engine and so for now I'm going to hope that someone else fixes this issue and I'm simply going to try to implement my workaround so that I can commit a version of my project that uses all of the latest nunjucks tech.

**Comment by carljm** [comment]
Sure, you're welcome to do that if you prefer. I wouldn't under-estimate your own qualifications, though -- you've already written the equivalent of a test for this in your above code, and the fix likely won't involve any of the hairy bits of "the insides of a template engine" -- it's really just a matter figuring out how to modify [this block of code](https://github.com/mozilla/nunjucks/blob/master/src/environment.js#L318-L323) so it handles defined properties, too.

**Comment by dschissler** [comment]
Challenge accepted.

**Comment by dschissler** [comment]
I haven't run this code yet but it seems correct in spirit.  The only issue that I can see is that it would possibly kill my dream of using a getter to alias to a sightly expensive in memory DB operation since it would always call on that property ahead of time, even when it wasn't being used in the template.

```javascript
function Context(ctx) {
  this.ctx = {};

  for(var k in ctx) {
    if(ctx.hasOwnProperty(k) || Object.getOwnPropertyDescriptor(ctx, k)) {
      this.ctx[k] = ctx[k];
    }
  }
}
```

**Comment by dschissler** [comment]
@carljm Also, I'm realizing that passing in the view instead of the model isn't such a bad idea, because then you can still access the getters on the model, totally fine and it doesn't triple memory usage because they would just be pointers.  The prototypes aren't appearing in the native `getOwnPropertyNames`, so it shouldn't be too much more inefficient.

What do you think?  Well, it works so I'm going to try changing the nunjucks source code and trying in my project.

Oh, and the downer is that `getOwnPropertyNames` isn't supported in IE8.  Which means that I guess the offending commit will need to be reverted or nunjucks will need to drop IE8 support.


```javascript
'use strict';
var nunjucks = require('nunjucks');

test1();
test2();

function test1(rawCtx) {
  var model, computedCtx;

  function Model() {
    this._modelValue = 'model_value';
    Object.defineProperty(this, 'value', {
      get: function() { return this._modelValue }
    });
  }
  Model.prototype.classFunction = function() {}
  model = new Model();

  computedCtx = computeContext(model);
  testComputedCtx(computedCtx);
}

function test2() {
  var model, computedCtx, html;

  model = {
    _modelValue: 'model_value',
    get value() {
      return this._modelValue;
    },
  }

  computedCtx = computeContext(model);
  testComputedCtx(computedCtx);
}

function computeContext(originalCtx) {
  var computedCtx = {};

  var props = Object.getOwnPropertyNames(originalCtx);
  for(var k in props) {
    var propertyName = props[k];
    if(originalCtx.hasOwnProperty(propertyName) || Object.getOwnPropertyDescriptor(originalCtx, propertyName)) {
      computedCtx[propertyName] = originalCtx[propertyName];
    }
  }

  return computedCtx;
}


function testComputedCtx(ctx) {
  var html;

  var count = 0;
  for (var k in ctx) {
    if (ctx.hasOwnProperty(k)) {
      ++count;
    }
  }

  if (count !== 2) {
    throw 'The context length should be exactly 2';
  }
  if (ctx['_modelValue'] !== 'model_value') {
    throw 'Direct property access failed.';
  }
  if (ctx['value'] !== 'model_value') {
    throw 'Getter property access failed.';
  }

  html = nunjucks.renderString('{{ _modelValue }}', ctx);
  if (html !== 'model_value') {
    throw 'Nunjucks failed to render correctly from a direct property access.';
  }
  console.log(html);

  html = nunjucks.renderString('{{ value }}', ctx);
  if (html !== 'model_value') {
    throw 'Nunjucks failed to render correctly from getter property access.';
  }
  console.log(html);
}
```

**Comment by carljm** [comment]
I don't think just reverting the commit that caused this problem is an option. That commit fixed a real bug that I think probably affected more people than this issue does. It might be worth exploring whether that bug could be fixed differently, in a way that doesn't require always copying the context. If possible, that would have the advantage of allowing getters to remain lazy.

I think we can use a fix using `getOwnPropertyNames` without totally dropping IE8 support, just by doing a support check -- if `getOwnPropertyNames` is not available, don't even try to support getters (since I doubt IE8 supports them anyway).

**Comment by dschissler** [comment]
Good idea and I can implement that in code.  Good news is that my fix works on the command line.  Unfortunately I just wasted an hour going nuts because I didn't realize that there are three versions of the source for different environments.  I thought that it was caching, webpack, everything.

So, how do I proceed to make a pull request with all of this?  I feel confident that I have the solution and now I need to know about tests and everything.  How do I get this committed so that I can check it off of my list.  I need to do this and another issue related to a long standing private fork of mine related to Phalcon Volt gettext parsing.

**Comment by carljm** [comment]
The only code that's relevant when working on nunjucks itself is the code in `src/` and the tests in `tests/`, and the fact that a run of `npm test` passes. All the different browser bundles in `browser/` are generated only at release time, and should be ignored when sending a pull request. Earlier in this thread I talked a bit about how to add a test to `tests/compiler.js` for this; should be pretty easy to translate your test code above into such a test.

**Comment by dschissler** [comment]
My fix, improves my limited test case but it doesn't seem to work with Ampersand in the wild.  I've resorted to the following in my view render code:

```javascript
      if (context.model) {
        context = {
          model: context.model
        }
      }
```

Funny enough my other limited fix doesn't seem to work with Ampersand.  I'm going to let it sit for a bit.  In my case this is not a bad fix.

**Comment by carljm** [comment]
Ok. Thanks for your exploration! We'll just leave this open - lots of useful information here if someone else comes along with motivation to fix this.
